orttraining/orttraining/training_ops/cuda/communication/recv.cc

```diff
@@ -35,6 +36,12 @@ Status Recv::ComputeInternal(OpKernelContext* ctx) const {
   const int64_t* remote_rank = remote_rank_tensor->template Data<int64_t>();
   const int src = static_cast<int>(*remote_rank);
 
+#ifndef NDEBUG
+  profile::NvtxNestedRangeCreator preRange(
+    "PreRecv-" + std::to_string(src), profile::Green);
+  preRange.Begin();
+#endif
```

Comment:
Would be nice to have auto-scope class and MACRO

Reply:
Thanks for sharing! For timing for `pre-send`, `send`, and `post-send`, using `Begin()` and `End()` here would be easier than using auto-scope. If we always consider a kernel as a whole, I totally agree that auto-scope is much better.

---
In reply to: [428780905](https://github.com/microsoft/onnxruntime/pull/4013#discussion_r428780905) [](ancestors = 428780905)

Reply:
You may use {} to limit the scope. Using macro would make adding new markers easier.

---
In reply to: [428807238](https://github.com/microsoft/onnxruntime/pull/4013#discussion_r428807238) [](ancestors = 428807238,428780905)

Reply:
I think using RAII here is a good idea too. It's possible to add an End() function to explicitly run the end logic prior at an earlier point. Not having to remember to call Begin() and End() in the presence of return statements and exceptions makes things simpler.

---
In reply to: [428883081](https://github.com/microsoft/onnxruntime/pull/4013#discussion_r428883081) [](ancestors = 428883081,428807238,428780905)

Reply:
There is a place we need this `Begin` and `End` when measuring `forward` and `backward` intervals --- 
```
#if !defined(NDEBUG) && defined(USE_CUDA) && !defined(_WIN32)
    if (node.Description() != "Backward pass" && !forward_range.IsBeginCalled()) {
      // Start timing forward pass when encountering the first forward node.
      forward_range.Begin();
    } else if (node.Description() == "Backward pass" && !backward_range.IsBeginCalled()) {
      // Start timing backward pass when encountering the first backward node.
      // In the meanwhile, forward range ends.
      forward_range.End();
      backward_range.Begin();
    }
#endif
```
Can this be supported using RAII? I am ok to create another RAII style class specially for `Wait` and `Record`. Sounds acceptable?

---
In reply to: [428992199](https://github.com/microsoft/onnxruntime/pull/4013#discussion_r428992199) [](ancestors = 428992199,428883081,428807238,428780905)

Reply:
@wschin Sorry for the late reply. Yes, I think it is possible to have an RAII class with explicit Begin() and End() methods for finer grained control. I'd imagine there could be an alternate constructor (along the lines of std::unique_lock(..., std::defer_lock_t)) which does not call Begin(). Having End() called automatically as needed would be nice.

I see the base class prints a warning if Begin() and End() are not called. That works but it's detected at runtime.

The PR is completed, but maybe consider this in the future. :)